tests/adjoint_2d_cylindrical/forward.py

```diff
@@ -0,0 +1,236 @@
+"""
+This runs the forward portion of the adjoint test case, to generate the reference
+final condition, and synthetic forcing (surface velocity observations).
+"""
+
+from gadopt import *
+import numpy as np
+
+
+def run_forward(visualise=False):
+    # Set up geometry:
+    geo_constants = get_geometry_parameters()
+    ref_values = get_reference_values()
+    # Start with a previously-initialised temperature field
+    with CheckpointFile("Checkpoint230.h5", mode="r") as f:
+        mesh = f.load_mesh("firedrake_default_extruded")
+        mesh.cartesian = False
+
+    # Set up function spaces for the Q2Q1 pair
+    V = VectorFunctionSpace(mesh, "CG", 2)  # Velocity function space (vector)
+    W = FunctionSpace(mesh, "CG", 1)  # Pressure function space (scalar)
+    DQ = FunctionSpace(mesh, "DQ", 2)  # Temperature function space (scalar, DQ2)
+    Q1 = FunctionSpace(mesh, "CG", 1)  # Average temperature function space (scalar, P1)
```

Comment:
I think it is usually a good idea to use `CG` and `DG` or `Q` and `DQ`, i.e. not relying on Firedrake doing some magic in the background. But that is also a personal opinion, so feel free to ignore me.